keg_auth/libs/authenticators.py

```diff
@@ -966,3 +975,137 @@ def get_authenticated_user(self):
+    the methods you wish to use.
+
+    Default settings are based on NIST guidelines and some common restrictions.
+    """
+
+    """
+    Minimum password length for check_length validation
+    """
+    min_length: int = 8
+
+    """
+    Character sets used for checking minimum "complexity" in check_character_set validation
+    """
+    required_char_types: typing.List[PasswordCharset] = [
+        PasswordCharset(_('lowercase letter'), string.ascii_lowercase),
+        PasswordCharset(_('uppercase letter'), string.ascii_uppercase),
+        PasswordCharset(_('number'), string.digits),
+        PasswordCharset(_('symbol'), ''.join(sorted(
+            set(string.printable)
+            - set(string.whitespace + string.ascii_letters + string.digits))
+        )),
+    ]
+
+    """
+    Minimum character number of different character types required in check_character_set validation
+    """
+    required_min_char_types: int = 3
+
+    def check_length(self, pw: str, user):
+        """
+        Raises PasswordPolicyError if a password is not at least min_length characters long.
+        :param pw: password to check
+        :param user: user entity
+        """
+        if len(pw) < self.min_length:
+            raise PasswordPolicyError(_(
+                'Password must be at least {min_length} characters long',
+                min_length=self.min_length
+            ))
+
+    def check_character_set(self, pw: str, user):
+        """
+        Raises PasswordPolicyError if a password does not contain at least one character from
+        at least `required_at_least_char_types` of the alphabets in `required_char_sets`.
+        :param pw: password to check
+        :param user: user entity
+        """
+
+        missing = []
+        for name, alphabet in self.required_char_types:
+            if not set(pw) & set(alphabet):
+                missing.append(name)
+
+        if len(missing) > len(self.required_char_types) - self.required_min_char_types:
+            if len(self.required_char_types) == 1:
+                message = _('Password must include a {type}', type=self.required_char_types[0].name)
+            else:
+                first_part = ', '.join(str(t.name) for t in self.required_char_types[:-1])
+                message = _(
+                    'Password must include at least {required} of {first} and/or {last}',
+                    required=self.required_min_char_types,
+                    first=first_part,
+                    last=self.required_char_types[-1].name
+                )
+
+            raise PasswordPolicyError(message)
+
+    def check_does_not_contain_username(self, pw: str, user):
+        """
+        Raises PasswordPolicyError if the password contains the username. This is case insensitive.
+        :param pw: password to check
+        :param user: user entity
+        """
+        try:
+            username = user.email.split('@')[0]
```

Comment:
This is technically incorrect, because it assumes that email is the username if email is available in the model. The rule of thumb in keg-auth is that the `get_username_key` model method will get you an attribute. Then, you can test that attribute on the user model to see if it is `EmailType`.